### PR TITLE
TELFE-85 feature(basic files): w/example cmds

### PR DESCRIPTION
**Pull request by ACoolmanTelicent**
💡  Ignore `./dist`

Essentially the initial commit

This CLI will be the base for our automation stuff.

# Walk through

## How to

https://github.com/telicent-oss/telicent-frontend-cli/assets/153610820/ac27bf7c-cb9a-48f0-829d-8d4d1424ed21

<details>
	<summary>More detail</summary>
	
	## Setup

https://github.com/telicent-oss/telicent-frontend-cli/assets/153610820/a03923ee-34ad-4940-b692-e9f1699c53ce

## Tour

https://github.com/telicent-oss/telicent-frontend-cli/assets/153610820/2c1c3d25-5bc6-4468-a901-d8eccc607b54

</details>

#  Future work

## _Auto check_ ideas

Once we install this CLI into a repo, this CLI can do things like:
- automatically add itself to lifecycle methods
- Compare repo's  config (e.g. lint/prettier) to the "blessed" standard
- Ensure PULL_REQUEST_TEMPLATE.md is correct
- Check folder structure
- Check OSS requirements
- autoupdate
- [Opt-in] Send system metrics (swapfile, CPU usage etc) to nominated web-hooks

## Handy command ideas

We can also add other useful, but manually invoked scripts e.g. 
`tefe system_report` - output versions of all the relevant FE binaries npm, nvm etc etc
or 
`tefe mvtmp ./node_modules` (instantly moves node_modules to `/tmp` folder with a unique name - where it will be wiped on system restart - so you can instantly do `npm install`)

Add the commit SHA and package version to the `./build/manifest.json` (or create `manifest.json`)

### CODE REVIEW

**Review by procolecato — COMMENTED** [review]
Thanks for inviting me to review this repo's documentation.

It's a good start and I think most of my comments are probably due to this being early days for the repo's documentation.  For example I've commented that usage should be more than just typing "TEFE help".  But if there are no commands to use right now then I can understand why this might be a placeholder.

Some of my comments might also be down to my inexperience with modern development techniques.  For example I wasn't sure how to apply TEFE to a particular frontend repo.

It might be helpful to state who the intended audience is, e.g. all developers of frontend components.  And would that be all frontend developers in the world or just Telicent developers?

File: LICENSE [hunk 1]
```diff
@@ -186,7 +186,7 @@
       same "printed page" as the copyright notice for easier
       identification within third-party archives.
 
-   Copyright [yyyy] [name of copyright owner]
+   Copyright 2024 Telicent Ltd.
```
[the remark this answers is not in the document]
Comment (ACoolmanTelicent) [reply LICENSE:189]: hahaha, I'm so reminded of [this scene from the Good Place™](https://www.youtube.com/watch?v=etJ6RmMPGko)

File: README.md [hunk 2]
```diff
@@ -1 +1,66 @@
-# telicent-frontend-cli
\ No newline at end of file
+## Telicent Frontend CLI ("TEFE" CLI)
+
+CLI utility help frontend repos stay aligned, and easy to work with.
```
[the remark this answers is not in the document]
Comment (ACoolmanTelicent) [reply README.md:3]: > How will it do this? Will it wrap-up and automate the scripts and checks described in the following paragraph?

Yep

> Is it used standalone as a CLI tool and / or is it linked into a package?

Yes both

File: README.md [hunk 3]
```diff
@@ -1 +1,66 @@
...
+
+Or to use the package globally:
+```sh
+yarn global link @telicent-oss/telicent-frontend-cli
+```
+</details>
+
+## Usage
+
+To list available commands
```
[the remark this answers is not in the document]
Comment (ACoolmanTelicent) [reply README.md:63]: Sure. Hm, Ok I'll probably pop the help strings directly into the readme.

Small Note: This project doesn't actually do anything yet

File: README.md [hunk 4]
```diff
@@ -1 +1,66 @@
...
+# IMPORTANT: If you use `yarn` to install locally, manually prefix with "yarn" e.g. `yarn tefe`
+```
+
+To test:
+```sh
+tefe version # or `yarn tefe version`
+```
+
+<details>
+  <summary>To use (and develop) commands locally:</summary>
```
[the remark this answers is not in the document]
Comment (procolecato) [reply README.md:27]: Ah, OK, thanks for the clarification I think that makes sense now.

You can use the TEFE CLI 'out of the box'.  This is what the wider README is describing.

OR 

If the standard TEFE CLI doesn't support what you want it to do you can develop new commands locally.  This is what the <details> section is describing.  As well as how to then use those new commands.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (procolecato) [reply README.md:63]: That's good to synchronise the help strings with the README!  

As mentioned, I did wonder if the emptiness of this section was simply due to the fact the project doesn't do anything yet.

**Comment by ACoolmanTelicent** [comment]
@procolecato  Did some quick tweaks 

Admittedly I didn't _carefully_ go through all input with a fine tooth comb, but maybe its enough

**Review by procolecato — APPROVED** [review]
That's great, Ash!  Given this is at an early stage of development of the CLI it's looking really good.

I like how you've given a bit more context around what it's for, who it's for and intended usage.

It's also good to see the link to the general process for developing CLI tools.

I think it's a good structure and starting point as the CLI develops.

As you've suggested on PRs for other repositories, I'll tweak the odd typos I've spotted myself in a separate PR.

**Comment by ACoolmanTelicent** [comment]
I regret not writing this in TS.

File: README.md [hunk 5]
```diff
@@ -1 +1,99 @@
...
+
+The general philosophy is: Automated alignment
+
+## Install
+
+```sh
+# Install within a package
+cd <npm package directory>;
+npm install @telicent-oss/telicent-frontend-cli
+# IMPORTANT: If you use `yarn` to install locally, manually prefix with "yarn" e.g. `yarn tefe`
```
Comment (dptelicent) [inline README.md:18]: As this is in the install section, shouldn't the suggestion if using Yarn be `yarn add @telicent-oss/telicent-frontend-cli`?

Comment (ACoolmanTelicent) [reply README.md:18]: Gonna presume we are a "yarn shop", we don't "support" npm

File: dist/autoupdate.d.ts [hunk 6]
```diff
@@ -0,0 +1,2 @@
+#!/usr/bin/env node
+export {};
```
Comment (ACoolmanTelicent) [inline dist/autoupdate.d.ts:2]: Ignore `./dist`

**Review by davec504 — CHANGES_REQUESTED** [review]
If you are adding the dist folder gitignore then the dist folder should be removed from this commit

**Comment by davec504** [comment]
> If you are adding the dist folder gitignore then the dist folder should be removed from this commit

After chat with @ACoolmanTelicent this is intentional, allowing the CLI to be node 16 compliant

File: src/info/utils/checkNpmContext.test.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      "process.cwd():,/mocked/path
+      Current dir:,
+        - is npm package:,true
+        - is @telicent-oss/telicent-frontend-cli:,true, (@telicent-oss/telicent-frontend-cli@123.345.456-test)
+        - has tefe installed:,
+        - has ./tefe.config.json:,{}"
+    `);
+  });
+
+  it.only("handles not being in an npm package", () => {
```
Comment (davec504) [inline src/info/utils/checkNpmContext.test.js:80]:
```suggestion
it("handles not being in an npm package", () => {
```

File: src/info/utils/checkNpmContext.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env node
+
+import * as fs from "fs";
```
Comment (davec504) [inline src/info/utils/checkNpmContext.ts:3]:
```suggestion
import fs from "fs";
```

File: src/autoupdate.ts [hunk 9]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env node
+import * as updateNotifier from "update-notifier";
+import * as fs from 'fs';
+
+const pkg = JSON.parse(`${fs.readFileSync('./package.json')}`);
+updateNotifier.default({ pkg }).notify();
```
Comment (davec504) [inline src/autoupdate.ts:6]:
```suggestion
import updateNotifier from "update-notifier";
import fs from 'fs';
const pkg = JSON.parse(`${fs.readFileSync('./package.json')}`);
updateNotifier({ pkg }).notify();
```

File: src/utils/formatJsonFile.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
+import * as util from "util";
+import * as fs from "fs";
```
Comment (davec504) [inline src/utils/formatJsonFile.ts:2]:
```suggestion
import util from "util";
import fs from "fs";
```

File: src/utils/tefe.config.json.utils.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
+import * as fs from "fs";
```
Comment (davec504) [inline src/utils/tefe.config.json.utils.ts:1]:
```suggestion
import fs from "fs";
```

**Review by davec504 — CHANGES_REQUESTED** [review]

**Review by AlecsMissangu — DISMISSED** [review]

File: README.md (hunk 5 above)
Comment (dptelicent) [reply README.md:18]:
```
# IMPORTANT: If you use `yarn` to install locally, manually prefix with "yarn" e.g. `yarn tefe`
```
👍🏻 Shouldn't the above comment be changed in that case, and maybe the default tefe command?

File: src/noise/index.ts [hunk 12]
```diff
@@ -0,0 +1,9 @@
+const noise:(...args: any[]) => void = (options) => {
```
Comment (davec504) [inline src/noise/index.ts:1]: Is this an easter egg or a template that has just been left in?

**Review by davec504 — APPROVED** [review]

File: src/noise/index.ts (hunk 12 above)
Comment (ACoolmanTelicent) [reply src/noise/index.ts:1]: I didn't mean to, but i figured I'll leave it in as a simple example for a while

**Review by AlecsMissangu — DISMISSED** [review]

File: src/utils/tefe.config.json.utils.test.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+      );
+    });
+  });
+
+  describe("createTefeJson", () => {
+    it("throws error if TEFE config already exists", () => {
+      fs.existsSync.mockReturnValue(true);
+
+      expect(() => createTefeJson()).toThrow(
+        `Can not create already existing: ${TEFE_CONFIG}`
```
Comment (dptelicent) [inline src/utils/tefe.config.json.utils.test.js:62]:
```suggestion
        `Cannot create already existing: ${TEFE_CONFIG}`
```

File: src/utils/tefe.config.json.utils.ts [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+      throw Error('Expected "version" (string) field in');
+    }
+    return true;
+  }
+  throw Error(`Perhaps file not created? could not find: ${TEFE_CONFIG}`);
+};
+
+export const createTefeJson = () => {
+  if (getTefeJson()) {
+    throw Error(`Can not create already existing: ${TEFE_CONFIG}`);
```
Comment (dptelicent) [inline src/utils/tefe.config.json.utils.ts:28]:
```suggestion
    throw Error(`Cannot create already existing: ${TEFE_CONFIG}`);
```

File: README.md [hunk 15]
```diff
@@ -1 +1,107 @@
...
+```sh
+# Install within a package
+cd <npm package directory>;
+yarn add @telicent-oss/telicent-frontend-cli
+# IMPORTANT: If you use `yarn` to install locally, manually prefix with "yarn" e.g. `yarn tefe`
+```
+
+To test:
+```sh
+tefe version # or `yarn tefe version`
```
[the remark this answers is not in the document]
Comment (dptelicent) [reply README.md:26]: **Optional suggestion**
As Yarn is the default.

**Review by dptelicent — DISMISSED** [review]

File: README.md (hunk 5 above)
Comment (dptelicent) [reply README.md:21]: See https://github.com/telicent-oss/telicent-frontend-cli/pull/1#discussion_r1580765385.

Comment (ACoolmanTelicent) [reply README.md]: [6ab1614](https://github.com/telicent-oss/telicent-frontend-cli/pull/1/commits/6ab1614456c68c6ca1ca1ea458aa8f9cdfd0534c)

**Review by davec504 — APPROVED** [review]

**Review by dptelicent — APPROVED** [review]

**Pull request merged**